### ISSUE TITLE
feat: Migrate `env.mjs` to `.ts`

### ISSUE THREAD

**Issue by chungweileong94** [t3-oss/create-t3-app#1369]
### Is your feature request related to a problem? Please describe.

It currently doesn't stop us from defining non-`NEXT_PUBLIC_` key as part of the client env.

### Describe the solution you'd like to see

If we migrate the `.env.mjs` to `.env.ts`, we can utilize the `satisfies` keyword to enforce the ``NEXT_PUBLIC_${string}`` as the client env key.
 
However, I'm actually pretty curious why `.mjs` for `env.mjs`? It would be good if someone can answer this🙏

I personally have the working solution in my personal project, and I am more than happy to help.

### Describe alternate solutions

-

**Comment by Pridestalkerr** [comment]
I am very curious to see your solution. Perhaps a link to a repository might encourage the assignment of the feat?

From my limited experience, the `env.mjs` is not a typescript file as it is loaded by the next app inside `next.config.mjs`. At the moment there seem to not be any plans to support `next.config.ts` by the **nextjs team**.
As to possible solutions I messed around with, `ts-import` could be used to load an `env.ts` file inside the config. There is also the option to pre-compile `env.ts` before nextjs starts, but that might be too messy.

**Comment by chungweileong94** [comment]
> From my limited experience, the env.mjs is not a typescript file as it is loaded by the next app inside next.config.mjs.

Right, I completely missed that one. I'm actually working on a vanilla NextJS template instead of create-t3-app template, and I basically borrowed the env implementation from here with some minor improvement, and thinking about contributing back.

Having to install extra tools for importing a single ts file into next.confit.js seems a bit too much IMO. But I just realized that JSDoc do support 'satisfies' now, so I think I can just port my implementation to js instead. Will try to create a PR once I get to my computer.
